File: java/src/main/java/com/example/cote/PGMRS/스킬_트리.java
package com.example.cote.PGMRS;

import java.util.*;
import java.util.stream.*;

public class 스킬_트리 {

  private static final int TEST_CASES = 10;
  private static final Random random = new Random(43);

  private static String generate(int limit) {
    int n = Math.min(random.nextInt(26 - 1) + 1, limit);
    Set<Character> set = new HashSet<>();

    while (set.size() < n) {
      set.add((char) (random.nextInt('Z' - 'A') + 'A'));
    }

    List<Character> chars = new ArrayList(set);

    Collections.shuffle(chars);

    return chars.stream().map(String::valueOf).collect(Collectors.joining());
  }

  private static int solve(String skill, List<String> skillTrees) {
    return (int)
        skillTrees.stream()
            .map(s -> s.replaceAll(String.format("[^%s]", skill), ""))
            .filter(skill::startsWith)
            .count();
  }

  public static void main(String... args) {

    for (int it = 0; it < TEST_CASES; it++) {
      String skill = generate(5);
      List<String> skillTrees =
          IntStream.range(1, random.nextInt(20))
              .mapToObj(i -> generate(26))
              .collect(Collectors.toList());
      System.out.println(solve(skill, skillTrees));
    }
  }
}

File: kotlin/src/main/kotlin/com/example/cote/PGMRS/스킬_트리.kt
package com.example.cote.PGMRS

import kotlin.random.*

val TEST_CASES = 10
val MAX_SKILLS = 5
val MAX_SKILL_TREES_LENGTH = 20
val random = Random(43)

fun generate(n: Int): String {

	val set = mutableSetOf<String>()

	do {
		val skill = ('A' + random.nextInt(26 - 1)).toString()
		set.add(skill)
	} while(set.size < n)

	return set.toMutableList().shuffled(random).joinToString("")
}

fun solve(skills: String, skillTrees: List<String>): Int {
	val regex = Regex(String.format("[^%s]", skills))
	
	return skillTrees.map{
		it.replace(regex, "")
	}.filter(skills::startsWith)
	.count()
}

fun main() {
	repeat(TEST_CASES) {
		val n = random.nextInt(MAX_SKILLS - 1) + 1
		val skills = generate(n)
		
		val skillTrees = IntRange(1, random.nextInt(MAX_SKILL_TREES_LENGTH - 2) + 2)
			.map {
				val k = random.nextInt(MAX_SKILLS - 1) + 1
				generate(k)
			}
		println(solve(skills, skillTrees))
	}
}